python/ray/autoscaler/_private/aws/node_provider.py

```diff
@@ -357,35 +357,35 @@ def _create_node(self, node_config, tags, count):
         # single SubnetId before invoking the AWS API.
         subnet_ids = conf.pop("SubnetIds")
 
+        # update config with min/max node counts and tag specs
+        conf.update({
+            "MinCount": 1,
+            "MaxCount": count,
+            "TagSpecifications": tag_specs
+        })
+
+        cli_logger_tags = {}
         for attempt in range(1, BOTO_CREATE_MAX_RETRIES + 1):
             try:
-                subnet_id = subnet_ids[self.subnet_idx % len(subnet_ids)]
-
-                self.subnet_idx += 1
-                conf.update({
-                    "MinCount": 1,
-                    "MaxCount": count,
-                    "NetworkInterfaces": [{
-                        "AssociatePublicIpAddress": True,
-                        "DeviceIndex": 0,
-                        "SubnetId": subnet_id,
-                    }],
-                    "TagSpecifications": tag_specs
-                })
-                _ = conf.pop("SubnetId", None)
-                security_group_ids = conf.pop("SecurityGroupIds", None)
-                if security_group_ids is not None:
-                    conf["NetworkInterfaces"][0]["Groups"] = security_group_ids
+                if "NetworkInterfaces" in conf:
```

Comment:
looks good, cc @xcharleslin

Reply:
> This looks good.
> 
> There's a validation error in test_autoscaler_yaml.py
> One way to fix it is to copy the lines about fillout_available_resources, replacing "Kubernetes" with "AWS" https://buildkite.com/ray-project/ray-builders-pr/builds/8970#a40ffc49-02f8-4875-8763-b9a67c9b99aa/216-356
> Another way is to add the resources field to each available node type in the new example config.
> 
> Also it would be great to add a test to test_autoscaler_aws checking the config logic.

Latest update has added support for AWS node provider unit test stubs, and uses them to exercise the happy path of this PR from config bootstrapping to node provisioning across all available node types in the example YAML.

I can follow up shortly with another small PR focused on exercising expected validation failures for things like conflict keys.

Reply:
I've also added empty resource fields to each node type in the example config to resolve the validation error in test_autoscaler_yaml.py.

Reply:
I'm also looking into the current failure of test_autoscaler_aws in buildkite...

Reply:
The latest update should fix the broken "test_network_interfaces" test of test_autoscaler_aws.py. The underlying EC2 client `describe_instances` call wasn't stubbed out at: https://github.com/ray-project/ray/blob/4971a9c7e6576c0a9fe10128f24f96f07244415c/python/ray/autoscaler/_private/aws/node_provider.py#L277

However, the test always succeeded locally because boto3 quietly fetched my local `~/.aws/credentials` file and completed the call directly against my AWS account.

I was also a little surprised to see that the EC2 client in question wasn't stubbed by the existing `ec2_client_stub` pytest fixture. It turns out that the `@lru_cache` decorator treats explicit specification of an optional default value differently than implicit specification of the same value. In this case, given a function definition like:
 ```python
 @lru_cache
 def resource_cache(name, region, max_retries=BOTO_MAX_RETRIES, **kwargs)
 ```
Calling `resource_cache("ec2", "us-west-2", BOTO_MAX_RETRIES)` resolves to a different cache key than `resource_cache("ec2", "us-west-2")`, even though `max_retries` equals `BOTO_MAX_RETRIES` in both cases. This is discussed a bit more over at: https://stackoverflow.com/questions/36771866/memoization-caching-with-default-optional-arguments.

The above detour, while interesting, made me think that it may be a good idea to open up a separate PR introducing a README to the AWS test directory discussing best practices for writing and running stubbed AWS unit tests. 🙂

Reply:
That lru_cache issue sounds pretty annoying...
Adding a README is a great idea!